src/plugins/intel_gpu/src/graph/layout_optimizer.cpp

```diff
@@ -1573,9 +1573,25 @@ impl_types layout_optimizer::get_preferred_impl_type(program_node& node, format
         if (node.is_type<fully_connected>()) {
             if (!is_node_for_onednn(node.as<fully_connected>()))
                 impl_candidate = impl_types::ocl;
-        } else {
-            if (node.is_dynamic()) {
-                impl_candidate = impl_types::ocl;
+        } else if (node.is_type<gemm>()) {
+            for (auto& fo : node.get_fused_primitives()) {
+                if (fo.is_type<eltwise>() && fo.typed_desc<eltwise>()->mode == eltwise_mode::sum) {
+                    if (onednn_add_fusing_helpers::get_add_fusing_type(node, fo) == add_fusing_type::binary_per_oc) {
+                        auto& data_node = node.get_dependency(fo.dep_start_idx);
+                        auto data_node_shape = data_node.get_output_layout().get_shape();
+                        auto gemm_out_shape = node.get_output_layout().get_shape();
+
+                        if (gemm_out_shape.size() >= 4) {
+                            auto onednn_gemm_batch_dim = gemm_out_shape[0] * gemm_out_shape[1];
+                            auto onednn_data_node_batch_dim = data_node_shape[0] * data_node_shape[1];
+
+                            // Fallback to impl_types::ocl for non-constant node to avoid broadcasting in runtime
+                            if (onednn_gemm_batch_dim != onednn_data_node_batch_dim && !data_node.is_constant()) {
+                                impl_candidate = impl_types::ocl;
```

Comment:
What Sergey implemented is:
* If the argument of binary_add is constant, add broadcast. It will be computed during constant propagation.
* If the argument of binary_add is not-constant, fallback to cldnn gemm.

Performance impact:
* cldnn gemm might be slower than onednn gemm on discrete graphics. So we need to check performance impact on dGPU.


We have several choices here:
1. For constant, add broadcast. For non-constant, use cldnn  (Sergey's choice)
** For non-constant case, we have not seen such model.
3. For constant, add broadcast, For non-constant, add broadcast, too.
4. For constant, add broadcast. For non-constant, do not fuse and use onednn for gemm.